8Knot/pages/contributions/visualizations/pr_first_response.py

```diff
@@ -0,0 +1,244 @@
+
+    # beginning to the end of time by the specified interval
+    dates = pd.date_range(start=earliest, end=latest, freq="D", inclusive="both")
+
+    # df for open prs and responded to prs in time interval
+    df_pr_responses = dates.to_frame(index=False, name="Date")
+
+    # every day, count the number of PRs that are open on that day and the number of
+    # those that were responded to within num_days of their opening
+    df_pr_responses["Open"], df_pr_responses["Response"] = zip(
+        *df_pr_responses.apply(
+            lambda row: get_open_response(df, row.Date, num_days),
+            axis=1,
+        )
+    )
+
+    df_pr_responses["Date"] = df_pr_responses["Date"].dt.strftime("%Y-%m-%d")
+
+    return df_pr_responses
+
+
+def create_figure(df: pd.DataFrame, num_days):
+
+    fig = go.Figure(
+        [
+            go.Scatter(
+                name="Prs Open",
+                x=df["Date"],
+                y=df["Open"],
+                mode="lines",
+                showlegend=True,
+                hovertemplate="PR's Open: %{y}<br>%{x|%b %d, %Y} <extra></extra>",
+                marker=dict(color=color_seq[1]),
+            ),
+            go.Scatter(
+                name="Response <" + str(num_days) + " days",
+                x=df["Date"],
+                y=df["Response"],
+                mode="lines",
+                showlegend=True,
+                hovertemplate="PRs: %{y}<br>%{x|%b %d, %Y} <extra></extra>",
+                marker=dict(color=color_seq[5]),
+            ),
+        ]
+    )
+    fig.update_layout(
+        xaxis_title="Time",
+        yaxis_title="Number of PRs",
+        font=dict(size=14),
+    )
+
+    return fig
+
+
+def get_open_response(df, date, num_days):
+    """
+    This function takes a date and determines how many
+    prs in that time interval are opened and if they have a response within num_days.
+
+    Args:
+    -----
+        df : Pandas Dataframe
+            Dataframe with pr assignment actions of the assignees
+
+        date : Datetime Timestamp
+            Timestamp of the date
+
+        num_days : int
+            number of days that a response should be within
+
+    Returns:
+    --------
+        int, int: Number of opened and responded to prs within num_days on the day
+    """
+    # drop rows that are more recent than the date limit
+    df_created = df[df["pr_created_at"] <= date]
+
+    # drops rows that have been closed after date
+    df_open = df_created[df_created["pr_closed_at"] > date]
+
+    # include prs that have not been close yet
+    df_open = pd.concat([df_open, df_created[df_created.pr_closed_at.isnull()]])
+
+    # column to hold date num_days after the pr_creation date for comparision
+    df_open["response_by"] = df_open["pr_created_at"] + pd.DateOffset(days=num_days)
+
+    # Inlcude only the prs that msg timestamp is before the responded by time
+    df_response = df_open[df_open["earliest_msg_timestamp"] < df_open["response_by"]]
```

Comment:
Just thought of this- what if the first message timestamp is from the PR-opener themselves? It seems common that the PR submitter could bump their PR or add details, which would be counted here, but maybe the contributor never receives a _community_ response?

Reply:
good point ill update it to also be that of a person that isnt the pr opener

Reply:
++